app/domain/authentication/authn_jwt/validate_and_decode/validate_and_decode_token.rb

```diff
@@ -0,0 +1,133 @@
+      # 1st we are validating only the signature.
+      # 2nd we are validating the claims, by checking the token content to decide which claims are mandatory
+      # for the 2nd validation
+      ValidateAndDecodeToken ||= CommandClass.new(
+        dependencies: {
+          verify_and_decode_token: ::Authentication::Jwt::VerifyAndDecodeToken.new,
+          fetch_jwt_claims_to_validate: ::Authentication::AuthnJwt::ValidateAndDecode::FetchJwtClaimsToValidate.new,
+          get_verification_option_by_jwt_claim: ::Authentication::AuthnJwt::ValidateAndDecode::GetVerificationOptionByJwtClaim.new,
+          logger: Rails.logger
+        },
+        inputs: %i[authentication_parameters fetch_signing_key]
+      ) do
+        extend(Forwardable)
+        def_delegators(:@authentication_parameters, :jwt_token)
+
+        def call
+          @logger.debug(LogMessages::Authentication::AuthnJwt::ValidatingToken.new)
+          validate_token_exists
+          fetch_signing_key
+          validate_signature
+          fetch_jwt_claims_to_validate
+          validate_claims
+          decoded_token_after_claims_validation
+          @logger.debug(LogMessages::Authentication::AuthnJwt::ValidatedToken.new)
+
+          @decoded_token_after_claims_validation
+        end
+
+        private
+
+        def validate_token_exists
+          raise Errors::Authentication::AuthnJwt::MissingToken if jwt_token.blank?
+        end
+
+        def fetch_signing_key(force_read: false)
+          @jwks = @fetch_signing_key.call(refresh: force_read)
+          @logger.debug(LogMessages::Authentication::AuthnJwt::SigningKeysFetchedFromCache.new)
+        end
+
+        def validate_signature
+          @logger.debug(LogMessages::Authentication::AuthnJwt::ValidatingTokenSignature.new)
+          ensure_keys_are_fresh
+          decoded_token_after_signature_only_validation
+          @logger.debug(LogMessages::Authentication::AuthnJwt::ValidatedTokenSignature.new)
+        end
+
+        def ensure_keys_are_fresh
+          decoded_token_after_signature_only_validation
+        rescue
+          @logger.debug(
+            LogMessages::Authentication::AuthnJwt::ValidateSigningKeysAreUpdated.new
+          )
+          # maybe failed due to keys rotation. Force cache to read it again
+          fetch_signing_key(force_read: true)
+        end
+
+        def decoded_token_after_signature_only_validation
+          @decoded_token_after_signature_only_validation ||= decoded_token(verification_options_for_signature_only)
+        end
+
+        def verification_options_for_signature_only
+          @verification_options_for_signature_only = {
+            algorithms: algorithms,
+            jwks: @jwks
+          }
+        end
+
+        def algorithms
+          @algorithms ||= SUPPORTED_ALGORITHMS
+        end
+
+        def decoded_token(verification_options)
+          @decoded_token = @verify_and_decode_token.call(
+            token_jwt: jwt_token,
```

Comment:
plesae choose between `jwt_token` and `token_jwt` (i prefer the first)

Reply:
if this requires a refactor that you don't want to do now then please create an issue and link it here.

Reply:
VerifyAndDecodeToken used by other authenticators so I don't want to change its name.
Im also prefer `token_jwt` but I don't want to touch other authenticators and don't want to change it in autun_jwt to `token_jwt`

Reply:
i'm fine with changing it to `jwt_token` later but please create an issue for it.